### PR TITLE
Serialize Service Mgmt API's report endpoint body correctly

### PR DESCRIPTION
**Pull request by mayorova**
This fixes the serialization issue with the `report` endpoint of the Service Management API, initially reported in this comment: https://github.com/3scale/porta/pull/3103#issuecomment-1413955116 (item 2).

Actually, it serializes differently in different cases:
- just using the single transaction element with the value from example:
  ```
  'service_token=token&service_id=123&transactions[user_key]=example&transactions[usage][hits]=1'
  ```
- the default element plus another element:
  ```
  'service_token=token&service_id=123&transactions[user_key]=example&transactions[usage][hits]=1&transactions=%7B%0A%20%20%22user_key%22%3A%20%22example%22%2C%0A%20%20%22usage%22%3A%20%7B%0A%20%20%20%20%22hits%22%3A%201%0A%20%20%7D%0A%7D'
  ```
- with manual changes in both elements:
  ```
  'service_token=token&service_id=123&transactions=%7B%0A%20%20%22user_key%22%3A%20%22example1%22%2C%0A%20%20%22usage%22%3A%20%7B%0A%20%20%20%20%22hits%22%3A%201%0A%20%20%7D%0A%7D&transactions=%7B%0A%20%20%22user_key%22%3A%20%22example2%22%2C%0A%20%20%22usage%22%3A%20%7B%0A%20%20%20%20%22hits%22%3A%201%0A%20%20%7D%0A%7D'
  ```

In order to fix this I considered several options:
1. Do it in `requestInterceptor`
   This was a bit messy, because at this point the request body is already serialized, and it would be tricky to parse, especially given that sometimes it's serialized as JSON, and other times as form-encoded string, and also could be mixed.

2. I also tried to write a plugin in the following way:
```
const ExecuteRequestPlugin: SwaggerUIPlugin = () => {
  return {
    statePlugins: {
      spec: {
        wrapActions: {
          executeRequest: (oriAction: (req: RequestData) => unknown) => (req: RequestData) => {
            // modify req.requestBody
            return oriAction(req)
          }
        }
      }
    }
  }
}
```
At first it looked promising, as `console.log(req)` was printing out the `requestBody` in the browser console, however I then realized that at this point in the execution `requestBody` is actually **not yet** available, and `console.log` was misleading.

3. So, I went with the approach of overriding the `fn.execute` function of the SwaggerUI, which in its turn is just imported from `swagger-js` library.

It's not ideal and quite "hacky", because it tweaks only the specific endpoint and in a very specific way.
Ideally, we could implement something more generic, that would serialize any object with config `deepObject` and `explode` in a way that we expect. Maybe, this would also potentially allow us to get rid of this hack of adding `[]` to the parameters names, like here:
https://github.com/3scale/porta/blob/eb07699ae413be4d6a408de16eb23c947bfc9c28/doc/active_docs/account_management_api.json#L2524-L2531

But I would leave this task for later, as an improvement.

For now, the serialization works as expected, the same way as in the original Swagger v1, with a slight difference on special characters encoding (e.g. spaces are encoded as `%20` rather than `+`. But I think it applies to all APIs in general, and should be OK.

**Example:**

*Before*
```
curl -v  -X POST \
  "https://su1.3scale.net/transactions.xml" \
  -d 'service_token=token&service_id=123&transactions%5B0%5D%5Buser_key%5D=key1&transactions%5B0%5D%5Btimestamp%5D=2011-12-30+22%3A15%3A31+%2B08%3A00&transactions%5B0%5D%5Busage%5D%5Bmetric1-1%5D=11&transactions%5B0%5D%5Busage%5D%5Bmetric1-2%5D=12&transactions%5B1%5D%5Buser_key%5D=key2&transactions%5B1%5D%5Btimestamp%5D=2011-12-30+22%3A15%3A31+-08%3A00&transactions%5B1%5D%5Busage%5D%5Bmetric2-1%5D=21&transactions%5B1%5D%5Busage%5D%5Bmetric2-2%5D=22'
```
decoded body:
```
'service_token=token&service_id=123&transactions[0][user_key]=key1&transactions[0][timestamp]=2011-12-30+22%3A15%3A31+%2B08%3A00&transactions[0][usage][metric1-1]=11&transactions[0][usage][metric1-2]=12&transactions[1][user_key]=key2&transactions[1][timestamp]=2011-12-30+22%3A15%3A31+-08%3A00&transactions[1][usage][metric2-1]=21&transactions[1][usage][metric2-2]=22'
```

*After*
```
curl -X 'POST' \
  'http://localhost:3001/transactions.xml' \
  -H 'accept: */*' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'service_token=token&service_id=123&transactions%5B0%5D%5Buser_key%5D=key1&transactions%5B0%5D%5Btimestamp%5D=2011-12-30%2022%3A15%3A31%20%2B08%3A00&transactions%5B0%5D%5Busage%5D%5Bmetric1-1%5D=11&transactions%5B0%5D%5Busage%5D%5Bmetric1-2%5D=12&transactions%5B1%5D%5Buser_key%5D=key2&transactions%5B1%5D%5Btimestamp%5D=2011-12-30%2022%3A15%3A31%20-08%3A00&transactions%5B1%5D%5Busage%5D%5Bmetric2-1%5D=21&transactions%5B1%5D%5Busage%5D%5Bmetric2-2%5D=22'
```

decoded body:
```
'service_token=token&service_id=123&transactions[0][user_key]=key1&transactions[0][timestamp]=2011-12-30 22%3A15%3A31 %2B08%3A00&transactions[0][usage][metric1-1]=11&transactions[0][usage][metric1-2]=12&transactions[1][user_key]=key2&transactions[1][timestamp]=2011-12-30 22%3A15%3A31 -08%3A00&transactions[1][usage][metric2-1]=21&transactions[1][usage][metric2-2]=22'
```

Before and after are exactly the same, except for the timestamp value:
```
2011-12-30+22%3A15%3A31+%2B08%3A00
vs
2011-12-30%2022%3A15%3A31%20%2B08%3A00
```

Need to check if the new encoding actually works with apisonator, it should as `%20` is standard.

The above bodies as screenshots in the UI:

**Before**
transactions-swagger1

**After** (it's missing a `,` after the timestamp field, but I don't have time to re-screenshot now)
transactions-openapi3

----

Docs on SwaggerUI plugins: https://github.com/swagger-api/swagger-ui/blob/master/docs/customization/plugin-api.md

### CODE REVIEW

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 1]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+    if (data && typeof data === 'object') {
+      Object.keys(data).forEach((key: string) => {
+        buildFormData(formData, data[key], parentKey ? `${parentKey}[${key}]` : key)
+      })
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
```
Comment (josemigallas) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:60]: Several issues with this method:
I always find it very confusing two mix up declaration order, object comes as an argument and it's should be the focus of this function but instead is used at the very end for the first time. `buildFormData` is complex enough to be declared elsewhere, in fact it can be extracted without refactoring since it does not depend on `objectToFormData`.

The fact that an empty object is being populated from scratch suggests to me a `reduce` could be used here, ideally:
```js
formData = body.reduce(buildFormData, {})
```

Recurrence is nice provided the method signature doesn't change but here both `data` and `parentKey` generate control couples which IMHO means it should be split (at least) in two. On top of that `formData` is empty during the first iteration colliding with the actual method signature:
```ts
Record<string, never>
// vs
Record<string, boolean | number | string>
```

There is not one test for this method. I've spent a fair amount of time looking at it and still cannot get the grasp of it so it will be hard to maintain. Considering its purpose is pretty straightforward I strongly suggest it be refactored. I'm working on an alternative right now, I really rather sacrifice verbosity for readability.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 2]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+    body.transactions = body.transactions.map(transaction => {
+      switch (typeof transaction) {
+        case 'object':
+          return transaction
+        case 'string':
+          try {
+            return JSON.parse(transaction) as BackendApiTransaction
+          } catch (error: unknown) {
+            return null
+          }
```
Comment (josemigallas) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:88]: `safeFromJsonString` catches errors and casts types:
```suggestion
          return safeFromJsonString<BackendApiTransaction>(transaction) ?? null
```
A question though, instead of filtering out `null` attributes, could we set `undefined` as default and it will be ignored without the need of a filter? Example:
```js
JSON.stringify({ a: undefined, b: 123 })
// -> "{\"b\":123}" 
```

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 3]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+        default:
+          return null
+      }
+    }).filter(element => element != null) as BackendApiTransaction[]
+  }
+  return objectToFormData(body as BodyValue)
```
Comment (josemigallas) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:94]: `body` is parsed to `BodyValue` even though it's the only place where `objectToFormData` is used. Why not using `BackendApiReportBody`?

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 4]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+const RequestBodyTransformerPlugin: SwaggerUIPlugin = () => {
+  return {
+    fn: {
+      execute: (req: ExecuteData): unknown => {
+        if (req.contextUrl.includes('api_docs/services/service_management_api.json')
+            && req.operationId === 'report'
+            && req.requestBody) {
+          req.requestBody = transformReportRequestBody(req.requestBody as BackendApiReportBody)
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
```
Comment (josemigallas) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:106]: This error will be fixed after typing `execute` in swagger.d.ts:
```ts
function execute (req: ExecuteData): unknown
```

File: app/javascript/src/Types/SwaggerTypes.ts [hunk 5]
```diff
@@ -17,3 +18,51 @@ export interface ApiDocsServices {
...
+  requestContentType: string;
+  requestInterceptor?: ((request: Request) => Promise<Request> | Request) | undefined;
+  responseContentType: string;
+  responseInterceptor?: ((response: Response) => Promise<Response> | Response) | undefined;
+  scheme: string;
+  securities: unknown;
+  server: string;
+  serverVariables: unknown;
+  spec: unknown;
+}
```
Comment (josemigallas) [inline app/javascript/src/Types/SwaggerTypes.ts:50]: The rest are utility types but I think this one strictly belongs to `swagger-client` and it should be defined in swagger.d.ts.

File: app/javascript/src/Types/swagger.d.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+declare module 'swagger-client/es/execute' {
+  export function execute (req: unknown): unknown
+}
```
Comment (josemigallas) [inline app/javascript/src/Types/swagger.d.ts:3]:
```suggestion
// Reference:
// - https://github.com/swagger-api/swagger-js/blob/master/src/index.js
// - https://github.com/swagger-api/swagger-js/blob/master/src/execute/index.js
declare module 'swagger-client' {
  import type { SupportedHTTPMethods } from 'swagger-ui'

  export interface ExecuteData {
    contextUrl: string;
    fetch: (arg: unknown) => unknown;
    method: SupportedHTTPMethods;
    operation: unknown;
    operationId: string;
    parameters: unknown;
    pathName: string;
    requestBody?: unknown;
    requestContentType: string;
    requestInterceptor?: ((request: Request) => Promise<Request> | Request) | undefined;
    responseContentType: string;
    responseInterceptor?: ((response: Response) => Promise<Response> | Response) | undefined;
    scheme: string;
    securities: unknown;
    server: string;
    serverVariables: unknown;
    spec: unknown;
  }
  function execute (req: ExecuteData): unknown
  export { execute }
}
```

File: package.json [hunk 7]
```diff
@@ -105,6 +105,7 @@
     "redux-thunk": "^2.3.0",
     "select2": "^4.0.6-rc.1",
     "swagger-ui": "^3.51.1",
+    "swagger-client": "^3.13.5",
```
Comment (josemigallas) [inline package.json:108]: Latest as today:
```suggestion
    "swagger-client": "3.18.5",
```

File: spec/javascripts/__mocks__/swagger-client/es/execute/index.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+    execute: jest.fn()
```
Comment (josemigallas) [inline spec/javascripts/__mocks__/swagger-client/es/execute/index.js:2]:
```suggestion
  execute: jest.fn()
```

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 9]
```diff
@@ -1,9 +1,12 @@
 import SwaggerUI from 'swagger-ui'
+// this is how SwaggerUI imports this function https://github.com/swagger-api/swagger-ui/pull/6208
+import { execute } from 'swagger-client/es/execute'
```
Comment (josemigallas) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:3]: If using the swagger.d.ts suggested above:
```suggestion
import { execute } from 'swagger-client'
```

**Review by josemigallas — CHANGES_REQUESTED** [review]

File: app/javascript/src/Types/swagger.d.ts (hunk 6 above)
Comment (jlledom) [inline app/javascript/src/Types/swagger.d.ts:3]: Why is this required and why is it under the `Types` directory?

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 10]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
   container.appendChild(div)
 }
 
+/**
+ * when using Record notation, the following error is thrown:
+ * 'TS2456: Type alias 'BodyValue' circularly references itself.'
+ */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+type BodyValue = boolean | number | string | { [key: string]: BodyValue }
```
Comment (jlledom) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:28]: I know not much of TS, I understand this means:

`BodyValue` is either a `boolean`, a `number`, a `string`, or an object with only  one item which key is a `string` and the value is a `BodyValue`.

Am I right?  I'm confused by the `[key: string]` syntax, is `key` a language keyword here?

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 11]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
+  const formData = {}
+  buildFormData(formData, object)
+  return formData
+}
```
Comment (jlledom) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:64]: I spent some time trying to understand this and I still don't understand:

1. How does this transforms an array into the `'an_array[0][first]': '1'` form? Are arrays being treated as objects?
2. How this URL encodes? where?

I think I need some comments here.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 1 above)
Comment (jlledom) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:60]: I think `buildFormData` is OK to be inside `objectToFormData`. Otherwise the only thing `objectToFormData` would do is to call `buildFormData`, no need to create a new function for that. In any case, what I would do is remove `objectToFormData` and call `buildFormData({}, body)` directly from `transformReportRequestBody`.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 12]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+
+/**
+ * Transforms the request body of the Service Management API Report, as
+ * SwaggerUI (or rather swagger-js) does not serialize arrays of objects properly
+ * The hack is to process the request body in two steps:
+ * 1. normalize the 'transactions' array, as its elements may be either an object (if the value is taken
+ *    from the example in the spec), or as a serialized JSON (if the field is changed manually)
+ * 2. "flatten" the objects by transforming them into form-data structure with the entries like
+ *    'transactions[0][app_id]': 'example'
+ *    'transactions[0][usage][hits]': 1
```
Comment (jlledom) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:74]: I checked with the debugger this function does the point 1, but I don't understand how it does the point 2, or what the point 2 really means, since this `body` is an object, not an string. I need some help.

**Review by jlledom — COMMENTED** [review]
Since the code is complicated, the first thing I did is try to make this work locally. I arranged everything (apisonator, etc) and tried.

It works fine and the backend registers the metrics, but I found a bug: it seems the number of hits registered by the backend is greater than I specified in the request. Please check the attached video, observe how the `Hits` metric is `25` and after sending `hits: 5` then `Hits` is increased to `36`, it should be `25+5=30`, which leads me to think it's adding up also the `6` I set for to the `hits.3` metric.

[Screencast from 2023-04-04 14-43-04.webm](https://user-images.githubusercontent.com/17052241/229855070-64ab193d-54e4-4534-9cc5-5318b1b27566.webm)

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 10 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:28]: I don't know TS either :grimacing: 

Your understanding is almost correct, except that `{ [key: string]: BodyValue }` means that BodyValue can be an object with **any** number (not only one) of items whose key is a `string` and the value is `BodyValue`. This is one way to define an object in TS, see https://www.typescriptlang.org/docs/handbook/2/objects.html#index-signatures
Another would be `Record<string, BodyValue>` (see https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type), but this was giving me an error as explained in the comments on lines 24-25.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 1 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:60]: > The fact that an empty object is being populated from scratch suggests to me a reduce could be used here

That's what I thought initially, but the problem is that we do need to traverse the whole object tree, so I am not sure that `reduce` would be that helpful (or easy to implement).

I'll address the rest of the comments later.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 11 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:64]: > 1. How does this transforms an array into the 'an_array[0][first]': '1' form? Are arrays being treated as objects?

Yes, the trick is in how JS treats arrays, basically it's like an object where the keys are the array indexes. So, Object.keys(myArray) returns ['0', '1', '2'] etc. See:
Screenshot from 2023-04-11 11-18-14

> 2. How this URL encodes? where?

I actually had some back-and-forth with this one, I tried applying `encodeURIComponent()` to the value (`formData[parentKey] = data ? encodeURIComponent(data) : ''`), but then I realised that in this case there was double-encoding (e.g. spaces looked like `%20%2520` or smth like that). I think, the original `execute` method of the `swagger-client` that we're calling from the plugin takes care of the encoding (I guess [these methods](https://github.com/swagger-api/swagger-js/blob/v3.15.0/src/http/index.js#L363-L405) are used).

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 12 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:74]: > I don't understand how it does the point 2, or what the point 2 really means, since this body is an object, not an string.

Does this test help understand what I mean by "flatten"? https://github.com/3scale/porta/pull/3294/files#diff-05c2318fa7df4d9d6736e8314e8684b020b655c0034ff6e692fe1ba3c48d28a6R5-R59

`body` indeed is an object, but the _leafs_ of the object tree are either number or strings (well, actually I defined the type of the body as `BodyValue = boolean | number | string | { [key: string]: BodyValue }`, but actually in this specific case `boolean` is not even used.

So, this is just the way we serialize the object body into array of `< string: number | string >` tuples.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 10 above)
Comment (jlledom) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:28]: I understand, so the `[]` mean and `Array`, thanks!

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 1 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:60]: > I always find it very confusing two mix up declaration order, object comes as an argument and it's should be the focus of this function but instead is used at the very end for the first time.

Well, the only reason why it is used in the end is because `buildFormData` is declared inside. I can extract it from there, no problem. It's just that I think thought of `objectToFormData` as a function that does everything, I think it's pretty useless if we just keep the last 3 lines (basically, I agree with what @jlledom says, except I don't think we need to call `buildFormData({}, body)` directly from `transformReportRequestBody`.
Anyway, I will see if I can make it more readable.

> `buildFormData` is complex enough to be declared elsewhere, in fact it can be extracted without refactoring since it does not depend on objectToFormData.

But do we need to have an exported function somewhere else (where?) if we know it is not going to be used by anything rather than this function? To me it makes more sense to keep the things together in this case, this will avoid the need to jump around multiple files/lines in the same file to understand what the function does.

> On top of that `formData` is empty during the first iteration colliding with the actual method signature: [...]

Doesn't `Record<string, boolean | number | string>` accept `{}` ?

> There is not one test for this method.

Fair enough, although the test would be for `objectToFormData` as it's the one that is exported.

> I've spent a fair amount of time looking at it and still cannot get the grasp of it so it will be hard to maintain.

What's confusing about it? What would you suggest?

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 2 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:88]: Nice fact about `JSON.stringify` trimming the undefined attributes...
But in this case we want an object, not a stringified JSON value. And I don't think stringifying and then parsing again is reasonable for the purpose of getting rid of empty values.

As an alternative, instead of the `.map().filter()` chain I can just do a simple loop populating a new array with non-empty objects only, maybe that would be less confusing and more efficient.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 3 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:94]: My idea was to make `objectToFormData` generic enough so it could potentially be reusable in case we need to do the same serialization for other endpoints - I believe it's not needed at this point, but IMO it doesn't make sense to make it more specific either.

And also in any case when we enter the recursive function we'd need to cast to a more generic class...

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 4 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:106]: OK, I'll try that

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 12 above)
Comment (jlledom) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:74]: OK, it's `objectToFormData` who actually does 2. It was my misunderstanding of `objectToFormData` that confused me here.

File: app/javascript/src/Types/swagger.d.ts (hunk 6 above)
Comment (mayorova) [reply app/javascript/src/Types/swagger.d.ts:3]: Well, that's basically types and modules definitions... Typescript requires this.

Comment (jlledom) [reply app/javascript/src/Types/swagger.d.ts:3]: But it's not this method already defined in the library?

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts [hunk 13]
```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
...
+  return {
+    fn: {
+      execute: (req: ExecuteData): unknown => {
+        if (req.contextUrl.includes('api_docs/services/service_management_api.json')
+            && req.operationId === 'report'
+            && req.requestBody) {
+          req.requestBody = transformReportRequestBody(req.requestBody as BackendApiReportBody)
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-call
+        return execute(req)
```
Comment (jlledom) [inline app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:107]: If I understand this correctly, this replaces the original execute method to the one defined here, and this one at the end calls the original one, is that right?

Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:107]: Yes, that's correct.

File: app/javascript/src/Types/swagger.d.ts (hunk 6 above)
Comment (mayorova) [reply app/javascript/src/Types/swagger.d.ts:3]: This doesn't work for me :shrug: 

Getting this error:

```
ERROR in ./app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts 120:15-23
"export 'execute' (imported as '_execute') was not found in 'swagger-client'
```
Screenshot from 2023-04-12 14-00-47

I would keep it as it is right now.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 1 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts]: I tried to give it a spin, but I did not manage to get rid of recursion in favor of `reduce` and friends.

I don't think it makes sense to waste more time on this. It's a very specific case, and it currently works for the issue it is meant to solve.

If you have an alternative implementation - please share :wink:

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 2 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts]: Removed the filtering and added `safeFromJsonString` in this commit https://github.com/3scale/porta/pull/3294/commits/5bd22380662cf65087c92006144ec8368c0d59a3#diff-78d4b14487564cb474b2638642cd81f8cbc2626033cfe769bb0fe1f6524af4f6R77

**Comment by mayorova** [comment]
@jlledom 
> which leads me to think it's adding up also the 6 I set for to the hits.3 metric.

Yes... that's the case. I don't like this behavior at all, and I raised an issue about that https://issues.redhat.com/browse/THREESCALE-7681
The last comment from Andreu points to another JIRA to make this behavior configurable.

But as of today this behavior is expected.

Thanks for taking time for verifying the end-to-end functionality! :raised_hands:

File: package.json (hunk 7 above)
Comment (josemigallas) [reply package.json:110]: Yes it would make sense 👍🏻

File: app/javascript/src/Types/swagger.d.ts (hunk 6 above)
Comment (josemigallas) [reply app/javascript/src/Types/swagger.d.ts:3]: I'll retry again, did you update the import to `import { execute } from 'swagger-client'`?

> I would keep it as it is right now.

If the types belong to a 3rd party library I would strongly recommend to move them into a `d.ts` file and import them from the pacakge, not from `Types/`.

Comment (mayorova) [reply app/javascript/src/Types/swagger.d.ts:3]: > did you update the import to `import { execute } from 'swagger-client'`?

Yes, I did.

I think the problem is that it needs to be consistent with the actual library code, and this is not how `execute` is actually exported in `swagger-client`, but I'm not sure if that's the reason.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 1 above)
Comment (josemigallas) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts]: I won't insist, I lost this battle. One last war cry in favor of SRP though: a function that _does everything_ is *wrong*. A function must do one and only one thing. And it's OK to have a 1 or 2 line function as long as it complies with SRP. It makes it more readable and more maintainable.

Comment (mayorova) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts]: LOL, "function that does everything"  was probably an overstatement :sweat_smile: IMO `objectToFormData` actually does 1 thing (transform one object format to another), but it's self-contained, so you don't need to jump around the code to understand what it does.

But well, I understand your point also. Does this look better to you? https://github.com/3scale/porta/pull/3294/commits/a7a2e450b11231ffed306df167dda82459d9b1d0

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 9 above)
Comment (josemigallas) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts:3]: OK this is wrong, the import has to be from 'swagger-client/es/execute'.

File: app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts (hunk 3 above)
Comment (josemigallas) [reply app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts]: I prefer naming interfaces with capital I, e.g. `IBodyValue` 🤓 BUT it's perfect like that 🙌🏻

**Review by josemigallas — APPROVED** [review]
